src/Symfony/Component/HttpKernel/EventListener/ValidationFailedExceptionListener.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\EventListener;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpKernel\Event\ExceptionEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+use Symfony\Component\Serializer\SerializerInterface;
+use Symfony\Component\Validator\Exception\ValidationFailedException;
+
+/**
+ * @author Konstantin Myakshin <molodchick@gmail.com>
+ */
+final class ValidationFailedExceptionListener implements EventSubscriberInterface
+{
+    private const SUPPORTED_FORMATS = [
+        'xml',
+        'json',
+    ];
+
+    public function __construct(private readonly SerializerInterface $serializer)
+    {
+    }
+
+    public function onKernelException(ExceptionEvent $event): void
+    {
+        $throwable = $event->getThrowable();
+        if (!$throwable instanceof ValidationFailedException) {
+            return;
+        }
+
+        $format = $event->getRequest()->getPreferredFormat('json');
+        if (!\in_array($format, self::SUPPORTED_FORMATS, true)) {
+            return;
+        }
+
+        $data = $this->serializer->serialize($throwable->getViolations(), $format);
+        $response = JsonResponse::fromJsonString($data, JsonResponse::HTTP_BAD_REQUEST);
```

Comment:
422?

Reply:
are we abusing JsonResponse for xml? :see_no_evil:

Reply:
this should indeed not use JsonResponse as different formats are supported

Reply:
changed to 422 + replaced `JsonResponse` with simple `Response`